Project 3

Mr-Powers-CPP opened the pull request.

Commit:
project 1 commit

shant-bekverdyan commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+    def __init__(self, player):
+        self.player = player
+        self.enemy = self.generate_enemy()
+
```
Add an enemy class to add different kinds of enemies, specific damage types, and characteristics.

class Enemy:
    def __init__(self, name, stats):
        self.name = name
        self.stats = stats

shant-bekverdyan commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+import random
+from player import Player
+
```
Consider adding a class for player stats. 

class Stats:
    def __init__(self, health, attack, defense)

Kimchi5000 commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+            print(f"💥 {self.player.name} lands a **CRITICAL STRIKE** for double damage!")
+            self.enemy["health"] -= damage  # Apply extra hit
+
+        if self.player.role == "Mage" and self.player.mana >= 20:
```
if self.player.role == "Mage" and self.player.mana >= 20:
            print(f"🔥 {self.player.name} casts **Fireball** on {self.enemy['name']}!")
            self.enemy["health"] -= 15
            self.player.mana -= 20  # Spend mana
 
You could consider adding a message for when the Mage role doesn't have enough mana to use their special ability

Suggested:
if self.player.role == "Mage" and self.player.mana >= 20:
            print(f"🔥 {self.player.name} casts **Fireball** on {self.enemy['name']}!")
            self.enemy["health"] -= 15
            self.player.mana -= 20  # Spend mana
else:
            print(f"{self.player.name} does not have enough mana")

Kimchi5000 commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+        print(f"🗡 {self.player.name} attacks! {self.enemy['name']} loses {damage} HP.")
+
+        # Special Abilities
+        if self.player.role == "Warrior" and random.randint(1, 100) <= 20:
```
defining critical strike chance, fireball damage, and mana cost as constants could be useful if you want to modify their values in the future.

Suggested:
CRITICAL_STRIKE_CHANCE = 20
FIREBALL_DAMAGE = 15
MANA_COST_FIREBALL = 20

                              .............
    
        # Special Abilities
        if self.player.role == "Warrior" and random.randint(1, 100) <= CRITICAL_STRIKE_CHANCE:
            print(f"💥 {self.player.name} lands a **CRITICAL STRIKE** for double damage!")
            self.enemy["health"] -= damage * 2

        if self.player.role == "Mage" and self.player.mana >= MANA_COST_FIREBALL:
            print(f"🔥 {self.player.name} casts **Fireball** on {self.enemy['name']}!")
            self.enemy["health"] -= FIREBALL_DAMAGE
            self.player.mana -= MANA_COST_FIREBALL

Kimchi5000 commented on Project/player.py:
```diff
@@ -11,23 +12,52 @@ def __init__(self, name, role):
-        print("💾 Player data saved!")
+# 🎭 Player Subclasses
+
+class Warrior(Player):
```
The attributes for each class could be moved into the player.json file and imported in.  The current player.json file doesn't seem to do much other than to show the example usage at the bottom.

Kimchi5000 commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+        self.player = player
+        self.enemy = self.generate_enemy()
+
+    def generate_enemy(self):
```
You could also consider creating a separate json file for the different enemy classes.  This would make access and modification easier in the future.

Kimchi5000 commented on Project/rules.py:
```diff
@@ -1,22 +1,72 @@
         except FileNotFoundError:
             print("⚠ No game rules file found. Using default settings.")
 
+    def apply_role_modifiers(self):
```
def apply_role_modifiers(self):
.....
self.xp_multiplier *= ....

There's an xp_multiplier for each class but I can't find anything that impelements this or handles experience gain in anyway.  You could add an experience threshold value in the player class to determine how much experience is needed to level up-
Suggested:
class Player:
    """Base class for player data (loads & saves from JSON)."""
    def __init__(self, name, role):
       self.experience_to_level_up = 100

Afterwards, maybe you could allow the player to gain a certain value for experience after killing an enemy.  
You might also have to determine what would happen after the character levels up(increase in health, mana, damage, etc.)

shervinatorX replied:
def gain_experience(self, amount):
    """
    Gain experience points and check for level up.
    """
    gained_xp = int(amount * self.xp_multiplier)
    self.experience += gained_xp
    print(f"⭐ {self.name} gained {gained_xp} XP!")

    if self.experience >= self.experience_to_level_up:
        self.level_up()

def level_up(self):
    """
    Level up the player and reset experience.
    """
    self.level += 1
    self.experience = 0
    self.experience_to_level_up += 50  # Optional: make it harder to level each time
    print(f"🎉 {self.name} leveled up to Level {self.level}!")

shant-bekverdyan commented on Project/main.py:
```diff
@@ -0,0 +1,69 @@
+        
+        choice = input("Enter your choice: ").strip()
+
+        if choice == "1":
```
Consider adding more functions to organize the code

handle_explore
handle_combat
handle_inventory
etc.

Kimchi5000 commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+        while self.player.health > 0 and self.enemy["health"] > 0:
+            self.player_attack()
+            if self.enemy["health"] <= 0:
+                print(f"🏆 {self.enemy['name']} defeated!")
```
In the rules.py file, I made a comment about the xp.multiplier... I think this could be a good place to implement experience gain.  Maybe something like...

        while self.player.health > 0 and self.enemy["health"] > 0:
            self.player_attack()
            if self.enemy["health"] <= 0:
                print(f"🏆 {self.enemy['name']} defeated!")
                self.player.gain_experience(10) 
                break

shant-bekverdyan commented on Project/player.py:
```diff
@@ -11,23 +12,52 @@ def __init__(self, name, role):
         self.stats = {"strength": 5, "agility": 3, "intelligence": 4}
         self.inventory = []
 
+    def save_profile(self, filename="data/player.json"):
```
You have two save_profiles. Line 15 and line 24. Possible combine them into one? Seems like duplicate code.

Jayden-gif-dev commented on data/game_rules.json:
I would move this into project folder, then move into and merge with the data folder for accessibility.
Currently it is somewhat disorganize

Jayden-gif-dev commented on data/player.json:
I would move this into project folder, then move into and merge with the data folder for accessibility.
Currently it is somewhat disorganize

Jayden-gif-dev commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+            {"name": "Orc", "health": 80, "attack": 15, "defense": 4, "evasion": 3},
+            {"name": "Dark Mage", "health": 60, "attack": 12, "defense": 3, "evasion": 7}
+        ]
+        return random.choice(enemies)
```
Consider using a class that you can import from a json file.
If you create a json file for it.
This could help with consistency with the `Player` class

Jayden-gif-dev commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+
+    def enemy_attack(self):
+        """Enemy's turn to attack."""
+        if random.randint(1, 100) <= self.player.stats.get("agility", 0) * 2:  # Higher agility = better dodge chance
```
If agility is 0 or missing, the enemy attack would always be successful, consider a default value e.g: 1 to ensure fairness

```suggestion
agility = max(self.player.stats.get("agility", 1), 1)
if random.randint(1, 100) <= agility * 2:

```

Jayden-gif-dev commented on Project/combat.py:
For this file I sees alot of emoji, that would somewhat clash with other command-prompt from other operators. I would suggest another way to express your game graphic.

Jayden-gif-dev commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+
+    def generate_enemy(self):
+        """Dynamically generates an enemy with randomized stats."""
+        enemies = [
```
I would move all the enemy data to the top as a constant for consistency if you don't want to move on with creating a json file for those data.

```suggestion
ENEMIES = [
    {"name": "Goblin", "health": 50, "attack": 10, "defense": 2, "evasion": 5},
    {"name": "Orc", "health": 80, "attack": 15, "defense": 4, "evasion": 3},
    {"name": "Dark Mage", "health": 60, "attack": 12, "defense": 3, "evasion": 7}
]

def generate_enemy(self):
    return random.choice(ENEMIES).copy()

```

Jayden-gif-dev commented on Project/combat.py:
```diff
@@ -0,0 +1,69 @@
+        final_damage = max(base_damage - enemy_defense, 1)  # Ensure at least 1 damage
+        return round(final_damage, 1)
+
+    def player_attack(self):
+        """Player's turn to attack."""
+        if random.randint(1, 100) <= self.enemy["evasion"]:  # Enemy dodges attack
+            print(f"💨 {self.enemy['name']} dodged the attack!")
+            return
```
For what you got so far with multiple classes.
I suggest using a role-specific methods inside each class (e.g `Warrior`, `Mage`).
And I would remove the emoji for compability issues.
```suggestion
def warrior_crit(self, damage):
    if random.randint(1, 100) <= 20:
        print(f"{self.player.name} lands a **CRITICAL STRIKE** for double damage!")
        self.enemy["health"] -= damage  # Extra hit

def mage_fireball(self):
    if self.player.mana >= 20:
        print(f"{self.player.name} casts **Fireball** on {self.enemy['name']}!")
        self.enemy["health"] -= 15
        self.player.mana -= 20

```

Jayden-gif-dev commented on Project/main.py:
```diff
@@ -0,0 +1,69 @@
+
+    while True:
+        choice = input("Enter 1, 2, or 3: ").strip()
+        name = input("Enter your character name: ").strip()
```
Right now your code would ask the name again on invalid class choice, which can be repetitive.
I would suggest move it outside the while loop.

```suggestion
name = input("Enter your character name: ").strip()

while True:
    choice = input("Enter 1, 2, or 3: ").strip()
    
    if choice == "1":
        return Warrior(name)
    elif choice == "2":
        return Mage(name)
    elif choice == "3":
        return Rogue(name)
    else:
        print("⚠ Invalid choice. Please enter 1, 2, or 3.")

```

Jayden-gif-dev commented on Project/main.py:
```diff
@@ -0,0 +1,69 @@
+    player.load_profile()
+
+    # Load game difficulty settings
+    rules = GameRules("Normal")  # Change to "Easy" or "Hard" as needed
```
You could let the user choose the difficulty from the start.

```suggestion
def choose_difficulty():
    print("\n🎚 Choose Difficulty: Easy, Normal, Hard")
    while True:
        difficulty = input("Enter difficulty: ").strip().capitalize()
        if difficulty in ["Easy", "Normal", "Hard"]:
            return difficulty
        print("⚠ Invalid input. Try again.")

rules = GameRules(choose_difficulty())

```

Jayden-gif-dev commented on Project/main.py:
```diff
@@ -0,0 +1,69 @@
+            print(f"🌍 {next(exploration)}")  # Show a random world event
+        elif choice == "2":
+            print("\n📜 Inventory:")
+            for item in player.inventory:
+                print(f"🔸 {item['name']} - {item['type']}")
```
I think this inventory could be a list of dictionaries if you change mose of your data into json file.
Assuming this is a item class
```
for item in player.inventory:
print(f"{item.name} - {item.type}")

```

Or

Assuming if your items are dictionaries, add a safe fallback
```
for item in player.inventory:
print(f"{item.get('name', 'Unknown')} - {item.get('type', 'Unknown')}")

```

Jayden-gif-dev commented on Project/main.py:
```diff
@@ -0,0 +1,69 @@
+            print("⚠ Invalid choice. Please try again.")
+
+if __name__ == "__main__":
+    main_menu()
```
For easier testing, you could consider adding a start_game() wrapper so it can be imported somewhere else easily.

```suggestion
def start_game():
    main_menu()

if __name__ == "__main__":
    start_game()

```

Jayden-gif-dev commented on Project/rules.py:
```diff
@@ -1,22 +1,72 @@
 class GameRules:
     """Loads game difficulty settings from JSON."""
-    def __init__(self, difficulty="Normal"):
+    def __init__(self, difficulty="Normal", player_role="Warrior"):
```
I have some problem with this might be might pathing of my OS. 
But I would suggest allowing the user to pass a custom path for flexibility and testability

```suggestion
def __init__(self, difficulty="Normal", player_role="Warrior", filename="data/game_rules.json"):
    self.difficulty = difficulty
    self.player_role = player_role
    self.filename = filename
    self.load_rules()

```

Jayden-gif-dev commented on Project/rules.py:
```diff
@@ -1,22 +1,72 @@
-            with open(filename, "r") as f:
+            with open(self.filename, "r") as f:
                 data = json.load(f)
                 self.__dict__.update(data.get(self.difficulty, {}))
```
Using `self.__dict__.update` is risky here since it will overwrite mothods/attributes accidentally. You can use explicit assignments instead if you desire to.

```suggestion
ruleset = data.get(self.difficulty, {})
self.enemy_damage_multiplier = ruleset.get("enemy_damage_multiplier", 1.0)
self.xp_multiplier = ruleset.get("xp_multiplier", 1.0)
self.item_drop_rate = ruleset.get("item_drop_rate", 1.0)
self.starting_gold = ruleset.get("starting_gold", 100)

```

Jayden-gif-dev commented on Project/rules.py:
```diff
@@ -1,22 +1,72 @@
+
 # Example Usage
-rules = GameRules("Hard")
+rules = GameRules("Hard", "Mage")  # Testing with a Mage
```
This should be remove. This could be ran under imports if not carefully protected.
You would only run examples under your main.py
Under
```
if __name__ == "__main__":
    rules = GameRules("Hard", "Mage")
    rules.load_rules()
```

Jayden-gif-dev commented on Project/world.py:
```diff
@@ -0,0 +1,12 @@
+    while True:
+        yield random.choice(events)
+
+# Example Usage
```
These type of example usage are very dangerous if you accidentally import it into your main file.
I would suggest only running these under your main.py
Under
```
if __name__ == "__main__":
```

shervinatorX committed:
Apply suggestions from code review

name = input("Enter your character name: ").strip()

while True:
    choice = input("Enter 1, 2, or 3: ").strip()

    if choice == "1":
        return Warrior(name)
    elif choice == "2":
        return Mage(name)
    elif choice == "3":
        return Rogue(name)
    else:
        print("⚠️ Invalid choice. Please enter 1, 2, or 3.")

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX committed:
Update Project/combat.py

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX committed:
Update Project/combat.py

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX committed:
Update Project/main.py

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX committed:
Update Project/main.py

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX committed:
Update Project/rules.py

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX committed:
Update Project/rules.py

Co-authored-by: Huynh Gia Phong Tat <112732639+Jayden-gif-dev@users.noreply.github.com>

shervinatorX commented on Project/rules.py:
```diff
@@ -1,22 +1,82 @@
-rules = GameRules("Hard")
+rules = GameRules("Hard", "Mage")  # Testing with a Mage
 rules.load_rules()
+
```
# Example Usage
if __name__ == "__main__":
    rules = GameRules("Hard", "Mage")  # Testing with a Mage
    rules.load_rules()

shervinatorX commented on Project/world.py:
```diff
@@ -0,0 +1,12 @@
+# Example Usage
+exploration = event_generator()
+for _ in range(3):
+    print(next(exploration))
```
# Example Usage
if __name__ == "__main__":
    exploration = event_generator()
    for _ in range(3):
        print(next(exploration))

Commit:
Project 3 commit
